---
title: "dnmt3a_tet2_mutated_week30"
author: "Sanketh"
date: "2024-01-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setting the directory and load required packages  

```{r}
rm(list=ls())
library(tidyverse)
library(Seurat)
library(ggplot2)
library(stringr)
#install.packages("Matrix")
#library(Matrix)
library(scCustomize)
library(harmony)
library(devtools)
library(presto)
library(DoubletFinder)
library(Matrix)
library(GMCM)
library(matrixStats)
library(SCENIC)
library(RcisTarget)
#devtools::install_github('immunogenomics/presto')
#install.packages("remotes")
#remotes::install_github("chris-mcginnis-ucsf/DoubletFinder")
setwd("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/")
getwd()
#Set the data dir to week 30 
dataDir <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/data/week30/")
rdaDir <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/data/week30/rda/")
imp_figs <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/imp_figs/")
```


```{r}
read_counts <- function(file){
  x <- readMM(file) %>%
  as.matrix()
  sample <- basename(file)
  sample <- str_replace(sample, "_matrix.*", "")
  genes_x <- read.table(file = paste0(dataDir,sample,"_genes.tsv.gz"), header = F, sep = "\t")
  genes <- genes_x[,2]
  #Be cautious while making the gene names unique 
  gene_names <- make.unique(genes) #This is a built in function in r to make the gene names unique 
  barcodes_x <- read.table(file = paste0(dataDir,sample,"_barcodes.tsv.gz"), header = F, sep = "\t")
  barcodes <- paste0(barcodes_x[,1],"_",sample)
  row.names(x) <- gene_names
  colnames(x) <- barcodes 
  return(x)
}
```

```{r}
read_barcodes <- function(file){
  y <- read.table(file, header = F, sep = "\t") %>%
    as.data.frame() %>%
    rename(Cellbarcode = "V1")
  sample <- basename(file)
  sample <- str_replace(sample,"_barcodes.tsv.gz","")
  barcode_metadata <- str_split(sample,pattern = "_") %>%
    unlist() %>%
    purrr::set_names(c("Sample_geo","week","mutation","type","subtype")) %>%
    as.list()
  barcode_metadata$mousetype <- paste(barcode_metadata$type,barcode_metadata$subtype,sep = ".")
  y$sample_geo <- rep(barcode_metadata$Sample_geo,nrow(y))
  y$week <- rep(barcode_metadata$week,nrow(y))
  y$genotype <- rep(paste(barcode_metadata$mutation,barcode_metadata$type,
                           barcode_metadata$subtype, sep="_"),nrow(y))
   if(barcode_metadata$mousetype == "45.1"){
     y$mutation <- rep("wt",nrow(y))
   }else{
     y$mutation <- rep(barcode_metadata$mutation,nrow(y))
    }
   return(y)
}
```



# Basic function to convert human to mouse gene names

```{r}
convert_human_to_mouse <- function(gene_list){
  
  output = c()
  
  for(gene in gene_list){
    class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name=="human"))[['DB.Class.Key']]
    if(!identical(class_key, integer(0)) ){
      mouse_genes = (mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name=="mouse, laboratory"))[,"Symbol"]
      for(mouse_gene in mouse_genes){
        output = append(output,mouse_gene)
      }
    }
  }
  
  return (output)
}
```

#Setting the cell cyle genes and converting the human to mouse genes  
  
```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
mouse_human_genes = read.csv("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt",sep="\t")
m.s.genes <- convert_human_to_mouse(cc.genes.updated.2019$s.genes)
m.g2m.genes <- convert_human_to_mouse(cc.genes.updated.2019$g2m.genes)
save.image(file = paste0(rdaDir,"cell_cycle_genes.rda"))
```


```{r}
#Reading in the count matrices
counts_files <- list.files(path = "./data/week30/", full.names = T, pattern = "*.mtx.gz")
samples <- map_chr(counts_files,basename)
samples <- str_replace(samples, "_matrix.*", "")
names(counts_files) <- samples
counts <- purrr::map(counts_files,read_counts) #Running now on all samples 
```

```{r}
##Write code to create metadata for the cells : 
barcode_files <- list.files (path = "./data/week30/", full.names = T, pattern = "*barcodes.tsv.gz")
names(barcode_files) <- samples
metadata_cells <- purrr::map(barcode_files,read_barcodes)
```

#Creating Seurat Object 
```{r}
library(Matrix)
objs <- purrr::map2(counts, metadata_cells,
                    ~CreateSeuratObject(counts = .x,
                                        meta.data = .y,
                                        )) #Can ignore the warning as Surat directly converts into sparse matrix which is req. 
```

```{r}
merged_seurat <- purrr::reduce(objs, function(x,y) {merge(x,y)})
#Once the merge is done the counts are not merged hence this needs to be 
#to merge the counts from different samples into one counts layer 
merged_seurat <- JoinLayers(merged_seurat)
merged_seurat
```

# Filter, normalize, scale, linear dimensionality reduction
```{r}
# Filter, normalize, scale, linear dimensionality reduction
merged_seurat[["percent.mt"]] <- PercentageFeatureSet(merged_seurat, pattern = "^mt-")
plot1 <- FeatureScatter(merged_seurat, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(merged_seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}
#First subset based on mitochondrial percentage : 
merged_seurat <- subset(merged_seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4400 & percent.mt < 15)
#Normalize the data 
merged_seurat <- Seurat::NormalizeData(merged_seurat, verbose = FALSE) 
```


```{r}
##Finding the most variable features and the most variable TFs present in the data 
##Next time replace with merged_seurat
Sds <- t(merged_seurat_harmony@assays$RNA$data) %>%
       as.matrix() %>%
       colSds(na.rm = T)

gene_var <- as_tibble((Sds))

features <- rownames(merged_seurat_harmony)

variance_genes <- cbind(features,gene_var) %>%
                  rename(variance = value)

variance_genes <- VariableFeatures(merged_seurat_harmony)

data(list = "motifAnnotations_mgi_v9" ,package = "RcisTarget")

tfs <- unique(motifAnnotations_mgi_v9$TF)

var_all_tfs <- variance_genes[variance_genes$features %in% tfs,]

ggplot(variance_genes) +
  aes(x = variance) +
  geom_histogram(bins = 30) +
  ylim(c(0,500)) +
  xlim(c(0.3,1.3)) + 
  theme_classic()

```


```{r}
ggplot(var_all_tfs) +
  aes(x = variance) +
  geom_histogram(bins = 100) +
  ylim(c(0,50)) +
  xlim(c(0.1,1.3)) + 
  theme_classic()
```


```{r}
var_genes <- variance_genes[(variance_genes$variance >= 0.5) | 
                            (variance_genes$features %in% tfs),]$features
#var_all_tfs[var_all_tfs$variance >= 0.3,]
```



```{r}
#Loading cell cycle genes
load(file = paste0(rdaDir,"cell_cycle_genes.rda"))
```




```{r}
merged_seurat <- subset(merged_seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4400 & percent.mt < 15)
merged_seurat <- Seurat::NormalizeData(merged_seurat, verbose = FALSE) %>% 
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>%   
  CellCycleScoring(g2m.features = m.g2m.genes,
                   s.features = m.s.genes,
                   set.ident = FALSE) %>% 
  ScaleData(vars.to.regress = c("S.Score", "G2M.Score")) %>% 
  RunPCA(pc.genes = merged_seurat@var.genes, npcs = 20, verbose = FALSE)

ElbowPlot(merged_seurat)
rm(counts,mouse_human_genes,barcode_files,
   counts_files,g2m.genes,m.g2m.genes,m.s.genes,samples,
   objs,s.genes,plot1,plot2)

save.image(file = paste0(rdaDir,"merged_seurat_day30_filtered.rda"))
###########Can directly start the workflow from here 
load(file = paste0(rdaDir,"merged_seurat_day30_filtered.rda"))
```

```{r}
# Run harmony integration
merged_seurat_harmony <- merged_seurat %>%
  RunHarmony("genotype")

ElbowPlot(merged_seurat_harmony)

merged_seurat_harmony <- RunUMAP(merged_seurat_harmony, reduction='harmony', 
                                 dims = 1:8)
merged_seurat_harmony <- FindNeighbors(merged_seurat_harmony, reduction='harmony', 
                                       dims = 1:8)
merged_seurat_harmony <- FindClusters(merged_seurat_harmony, resolution = 0.7)
```


```{r}
# Doublet detection and removal
sweep.list <- paramSweep(merged_seurat_harmony, PCs = 1:16)
sweep.stats <- summarizeSweep(sweep.list)
bcmvn <- find.pK(sweep.stats)
nExp <- round(0.03*nrow(merged_seurat_harmony@meta.data)) 
merged_seurat_doublets <- doubletFinder(merged_seurat_harmony, PCs = 1:8, pN = 0.25, pK = 0.09, nExp = nExp, reuse.pANN = FALSE, sct = FALSE)
merged_seurat_harmony = merged_seurat_doublets[, merged_seurat_doublets@meta.data[, DF.classifications_0.25_0.09_510] == "Singlet"]
merged_seurat_harmony = subset(x = merged_seurat_doublets, subset = DF.classifications_0.25_0.09_510 == "Singlet")
rm(merged_seurat)
save.image(file = paste0(rdaDir,"merged_seurat_harmony.rda"))
dataDir <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/data/week30/")
rdaDir <- ("/Users/sanketh/Desktop/proj1_tet_dnmt3a_mut/data/week30/rda/")
```

```{r}
load(file = paste0(rdaDir,"merged_seurat_harmony.rda"))
#merged_seurat_harmony@meta.data$genotype_clusters <- cellInfo$genotype_clusters
## Modify the meatadata column because muutation should be changed : All Cd45.1 should be wt and cd45d.2 should be the various mutations 
cellInfo$mutation[cellInfo$genotype %in% c('dnmt3a_45_1','tet2_45_1')] <- "wt"
cellInfo$genotype_clusters <- paste(cellInfo$mutation,cellInfo$seurat_clusters,sep = "_")
merged_seurat_harmony@meta.data <- cellInfo
cellInfo$genotype_clusters_ind <- paste(cellInfo$genotype,cellInfo$seurat_clusters,sep = "_")
merged_seurat_harmony@meta.data <- cellInfo
merged_seurat_harmony$seurat_clusters <- as.factor(merged_seurat_harmony$seurat_clusters) %>%
                                         as.character()
merged_seurat_harmony$genotype_clusters <- as.factor(merged_seurat_harmony$genotype_clusters) %>%
                                           as.character()
merged_seurat_harmony$genotype_clusters_ind <- as.factor(merged_seurat_harmony$genotype_clusters_ind) %>%
                                            as.character()
```


```{r}
DimPlot(merged_seurat_harmony, reduction = "umap", label = T,label.size = 2)
#ggsave(plot = umap_celltypes, filename = paste(imp_figs,"umap_celltypes.pdf"),dpi = 50)
```




```{r}
umap_genotype <- DimPlot(merged_seurat_harmony, reduction = "umap", label = F, group.by = "genotype")
ggsave(plot = umap_genotype, filename = paste(imp_figs,"umap_genotype.pdf"), dpi = 50)
```

```{r}
DimPlot(merged_seurat_harmony, reduction = "umap", label = F, group.by = "genotype_clusters")
```



```{r}
table(Idents(merged_seurat_harmony))
```



##FindMarkers : Find the differentially expressed genes in each of the different clusters to 
#annotate them :

```{r}
# Find markers to enable cluster annotation (supplementary file)
markers <- FindAllMarkers(merged_seurat_harmony, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC) -> top10
markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC) -> top20
write.csv(top10, file = "top10.csv")
```



```{r}
dist_genotype <- merged_seurat_harmony@meta.data %>%
                 group_by(seurat_clusters, mutation) %>%
                 summarise(n = n())
table(merged_seurat_harmony@meta.data$mutation)
table(merged_seurat_harmony@meta.data$seurat_clusters)
```

```{r}
ggplot(dist_genotype) +
  aes(x = seurat_clusters,
      y = n, fill = mutation) +
  geom_bar(stat = "identity",
           position = "fill") + 
  coord_flip() +
  scale_fill_brewer(palette = 3) +
  theme_classic()
#ggsave(filename = "imp_figs/cell_distributions.pdf")
```

```{r}
dist_cellphase <- merged_seurat_harmony@meta.data %>%
                 group_by(seurat_clusters, Phase) %>%
                 summarise(n = n())
```

```{r}
ggplot(dist_cellphase) +
  aes(x = seurat_clusters,
      y = n, fill = Phase) +
  geom_bar(stat = "identity",
           position = "fill") + 
  scale_fill_brewer(palette = 1) +
  coord_flip() +
  theme_classic()
ggsave(filename = "./imp_figs/cellcycle_distribution.pdf")
```

##Renaming the idents based on the expression of some known markers 
```{r}
merged_seurat_harmony <- RenameIdents(object = merged_seurat_harmony,
                                      '0' = "B cells", '1' = "T cells", 
                                      '2' = "Resident-like Macrophages", 
                                      '3' = "Inflamatory Macrophages", '4' = "Neutrophils",
                                      '5' = "TRM2hi Macrophages", '6' = "T cells",
                                      '7' = "Monocytes", '8' = "CXCR6+ T cells",
                                      '9' = "Dendritic cells", '10' = "Mixed lymphocytes",
                                      '11' = "Mixed Lymphocytes", '12' = "B cells", 
                                      '13' = "Proliferating T cells", '14' = "Proliferating T cells",
                                      '15' = "B cells", '16' = "Vascular Fibroblasts"
                                      )
table(Idents(merged_seurat_harmony))
```



```{r}
#cellInfo <- merged_seurat_harmony@meta.data
#cellInfo$seurat_clusters <- Idents(merged_seurat_harmony)
#cellInfo$genotype_clusters <- paste(cellInfo$mutation,cellInfo$seurat_clusters,sep = "_")
saveRDS(object = cellInfo, file = "./data/week30/rda/cellInfo.rds")
#rm(cellInfo)
#cellInfo <- load(file = "./data/week30/rda/cellInfo.rds")
```


```{r}
#merged_seurat_harmony@meta.data$seurat_clusters <- Idents(merged_seurat_harmony)
dist_genotype_ann <- merged_seurat_harmony@meta.data %>%
                 group_by(seurat_clusters,mutation) %>%
                 summarise(n = n())
```

```{r}
#Just plot proportions and not whole numbers !!!
ggplot(dist_genotype_ann) +
  aes(x = seurat_clusters,
      y = n,
      fill = mutation) +
  geom_bar(stat = "identity") + 
  #coord_flip()+
  scale_fill_brewer(palette = 3) +
  theme_classic()
```






```{r}
cellInfo$genotype_clusters <- as.factor(cellInfo$genotype_clusters)
cellInfo$genotype_clusters
```


```{r}
##Finding the most variable features 

Sds <- t(merged_seurat_harmony@assays$RNA$data) %>%
       as.matrix() %>%
       colSds(na.rm = T)

gene_var <- as_tibble((Sds))

features <- rownames(merged_seurat_harmony)

variance_genes <- cbind(features,gene_var) %>%
                  rename(variance = value)


data(list = "motifAnnotations_mgi_v9" ,package = "RcisTarget")

tfs <- unique(motifAnnotations_mgi_v9$TF)

var_all_tfs <- variance_genes[variance_genes$features %in% tfs,]

ggplot(variance_genes) +
  aes(x = variance) +
  geom_histogram(bins = 200) +
  ylim(c(0,500)) +
  xlim(c(0,1.3)) + 
  theme_classic()

variance_genes[(variance_genes$variance >= 0.5) | (variance_genes$features %in% tfs),]
var_all_tfs[var_all_tfs$variance >= 0.3,]

```


```{r}
ggplot(var_all_tfs) +
  aes(x = variance) +
  geom_histogram(bins = 200) +
  ylim(c(0,70)) +
  xlim(c(0,1)) + 
  theme_classic()
```


```{r}
var_genes <- variance_genes[variance_genes$variance >= 0.5 | 
                              variance_genes$features %in% tfs,]$features
```



```{r}
FeaturePlot(merged_seurat_harmony, features = c("Xbp1","Cepbp"), split.by = "mutation")
```



```{r}
DimPlot(merged_seurat_harmony, reduction = "umap", split.by = "genotype")
```


```{r}
DimPlot(merged_seurat_harmony, reduction = "umap", split.by = "mutation")
```



```{r}
FeaturePlot(merged_seurat_harmony, features = c ("Maf"), split.by = "mutation")
```


```{r}
FeaturePlot(merged_seurat_harmony, features = c ("Jun"), split.by = "mutation")
```






```{r}
# Ridge plots - from ggridges. Visualize single cell expression distributions in each cluster
features = "Cd19"
RidgePlot(merged_seurat_harmony, features = features, ncol = 2)
```





```{r}
VlnPlot(merged_seurat_harmony, features = features)
```


```{r}
features = "Jun"
VlnPlot(merged_seurat_harmony, features = features, split.by = "mutation")
```


```{r}
features = "Trem2"
VlnPlot(merged_seurat_harmony, features = features)
```

```{r}
features = "Spp1"
VlnPlot(merged_seurat_harmony, features = features)
```


```{r}
features = "Maf"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Xbp1"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Maf"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Cebpb"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Nlrp3"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```



```{r}
features = "Fosb"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Jun"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```



```{r}
features = "Atf3"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```


```{r}
features = "Gata3"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```

```{r}
features = "Klf4"
VlnPlot(merged_seurat_harmony, features = features,split.by = "mutation")
```






##Subsetting seurat objects into 5 different objects based on their 
#genotype to run SCENIC to see if there are any potential differnces 
```{r}
rm(bcmvn,merged_seurat_doublets,merged_seurat_harmony,metadata_cells,sweep.stats,sweep.list)
save.image(paste0(rdaDir,"seurat_objs_genotypes_noidents.rda")) #Can directly load this into the scenic pipeline to run scenic
```

```{r}
#Idents(dnmt3a_45_1_seurat)
```
```{r}
table(merged_seurat_harmony@meta.data$genotype_clusters)
```

```{r}
table(merged_seurat_harmony@meta.data$seurat_clusters)
```

```{r}
table(merged_seurat_harmony@meta.data$genotype_clusters_ind)
```

##Subsetting the seurat object such that it has all the highly variable genes and all the TFs 
```{r}
variance_genes <- VariableFeatures(merged_seurat_harmony)
data(list = "motifAnnotations_mgi_v9" ,package = "RcisTarget")
tfs <- unique(motifAnnotations_mgi_v9$TF)
tfs <- setdiff(tfs,variance_genes)
var_genes <- c(variance_genes,tfs)
merged_seurat_harmony <- subset(merged_seurat_harmony,features = var_genes)
dim(merged_seurat_harmony)
```
```{r}
merged_seurat_harmony@assays
```


##Convert seurat object to h5ad object that can be imported directly into python
```{r}
##Convert seurat object to anndata object 
library(Seurat)
library(SeuratDisk)
#Convert into s3 object 
merged_seurat_harmony[["RNA"]] <- as(merged_seurat_harmony[["RNA"]], "Assay")
SaveH5Seurat(merged_seurat_harmony, filename = "merged_seurat.h5Seurat")
Convert("merged_seurat.h5Seurat", dest = "h5ad")
```



```{r}
library(reticulate)
#devtools::install_github("cellgeni/sceasy")
library(sceasy)
use_condaenv('pyscenic_env')
merged_seurat_harmony[["RNA"]] <- as(merged_seurat_harmony[["RNA"]], "Assay")
sceasy::convertFormat(merged_seurat_harmony, from="seurat", to="anndata",
                       outFile='merged_seurat.h5ad')
```
```{r}
check_genes <- VariableFeatures(merged_seurat_doublets)
```




```{r}
library(SCP)
adata <- srt_to_adata(merged_seurat_harmony)
adata$write_h5ad("merged_seurat.h5ad")
```


